internal/controller/sriov_controller.go

```diff
@@ -0,0 +1,287 @@
+func (r *SRIOVReconciler) generateSRIOVNetworkNodePolicy(railName, deviceName string, mtu uint) *sriovnetworkv1.SriovNetworkNodePolicy {
+	nicSelector := &sriovnetworkv1.SriovNetworkNicSelector{
+		PfNames: []string{deviceName},
+	}
+	bridge := &sriovnetworkv1.Bridge{
+		OVS: &sriovnetworkv1.OVSConfig{
+			Bridge: sriovnetworkv1.OVSBridgeConfig{
+				DatapathType: "netdev",
+				OtherConfig:  map[string]string{"dpdk-init": "true"},
+			},
+			Uplink: sriovnetworkv1.OVSUplinkConfig{
+				Interface: sriovnetworkv1.OVSInterfaceConfig{
+					Type: ovsNetworkInterfaceType,
+				},
+			},
+		},
+	}
+	nodeState := &sriovnetworkv1.SriovNetworkNodePolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      railName,
+			Namespace: r.SriovObjNamespace,
+			Labels: map[string]string{
+				spectrumXSRIOVControllerName: "",
+			},
+		},
+		// NOTE: Temporary ignore MTU type conversion until we address it in SR-IOV Network Operator
+		//nolint:gosec
+		Spec: sriovnetworkv1.SriovNetworkNodePolicySpec{
+			ResourceName: railName,
+			Mtu:          int(mtu),
+			NumVfs:       1,
+			NicSelector:  *nicSelector,
+			IsRdma:       true,
+			EswitchMode:  "switchdev",
+			Bridge:       *bridge,
+		},
+	}
+
+	nodeState.ObjectMeta.ManagedFields = nil
+	nodeState.SetGroupVersionKind(sriovnetworkv1.GroupVersion.WithKind(sriovNodePolicyType))
+	return nodeState
+}
+
+// reconcileDelete delete created OVS Network remove finalizer
+func (r *SRIOVReconciler) reconcileDelete(ctx context.Context, cm *corev1.ConfigMap) (ctrl.Result, error) {
+	logr := log.FromContext(ctx)
+	logr.Info("Reconciling delete")
+	if err := r.deleteOVSNetworks(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+	if err := r.deleteSRIOVNetworkNodePolicies(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+	logr.Info("Removing finalizer")
+	controllerutil.RemoveFinalizer(cm, spectrumXSRIOVFinalizer)
+	cm.ObjectMeta.ManagedFields = nil
+	if err := r.Client.Patch(ctx, cm, client.Apply, client.ForceOwnership, client.FieldOwner(spectrumXSRIOVControllerName)); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error while removing finalizer to configmap %w", err)
+	}
+	return ctrl.Result{}, nil
+}
+
+// deleteSRIOVNetworkNodePolocies deletes all the SRIOVNetworkNodePolocies created by controller
+func (r *SRIOVReconciler) deleteSRIOVNetworkNodePolicies(ctx context.Context) error {
+	p := &unstructured.Unstructured{}
+	p.SetGroupVersionKind(sriovnetworkv1.GroupVersion.WithKind(sriovNodePolicyType))
+	if err := r.Client.DeleteAllOf(ctx, p, client.InNamespace(r.SriovObjNamespace), client.MatchingLabels{
+		spectrumXSRIOVControllerName: "",
+	}); err != nil {
+		return fmt.Errorf("error while removing all %s: %w", p.GetObjectKind().GroupVersionKind().String(), err)
+	}
+	// Wait for all instances to be deleted
+	timeout := 30 * time.Second
+	interval := 1 * time.Second
+	return wait.PollUntilContextTimeout(ctx, interval, timeout, true, func(ctx context.Context) (bool, error) {
```

Comment:
usually it's bad practices to hold a controller reconcile loop. it's better to delete and check again with a reconcile

Reply:
We did the same for IPAM objects here https://github.com/Mellanox/spectrum-x-operator/blob/main/internal/controller/ipam_controller.go#L210

We can fix in a follow up in both